lib/Sema/CSSimplify.cpp

```diff
@@ -1560,25 +1560,32 @@ isConversionAllowedBetween(FunctionTypeRepresentation rep1,
 
 // Returns 'false' (i.e. no error) if it is legal to match functions with the
 // corresponding function type representations and the given match kind.
-static bool matchFunctionRepresentations(FunctionTypeRepresentation rep1,
-                                         FunctionTypeRepresentation rep2,
+static bool matchFunctionRepresentations(FunctionType::ExtInfo einfo1,
+                                         FunctionType::ExtInfo einfo2,
                                          ConstraintKind kind,
+                                         ConstraintSystemOptions options,
                                          ConstraintLocatorBuilder locator) {
+  auto rep1 = einfo1.getRepresentation();
+  auto rep2 = einfo2.getRepresentation();
+  bool clangTypeMismatch =
+      (options.contains(ConstraintSystemFlags::UseClangFunctionTypes) &&
+       (einfo1.getClangTypeInfo() != einfo2.getClangTypeInfo()));
   switch (kind) {
   case ConstraintKind::Bind:
   case ConstraintKind::BindParam:
   case ConstraintKind::BindToPointerType:
   case ConstraintKind::Equal:
-    return rep1 != rep2;
-      
+    return (rep1 != rep2) || clangTypeMismatch;
+
   case ConstraintKind::Subtype: {
     auto last = locator.last();
     if (!(last && last->is<LocatorPathElt::FunctionArgument>()))
       return false;
     
     // Inverting the result because matchFunctionRepresentations
     // returns false in conversions are allowed.
-    return !isConversionAllowedBetween(rep1, rep2);
+    return !isConversionAllowedBetween(rep1, rep2) ||
+           ((rep1 == rep2) && clangTypeMismatch);
```

Comment:
Could you clarify what is the expected subtyping rule here, In particular why it's okay to have different reprs even in a presence of an error - `false` here means - representations match.

Reply:
The subtyping rule here (for now) is that converting between `@convention(c)` function types with different Clang types is not okay (they have to match exactly). Existing behavior should be unchanged. Here is the breakdown of cases:

1. `rep1 == rep2`: In this case, `isConversionAllowedBetween(rep1, rep2) == true`. If `clangTypeMismatch == true`, then this function will return true, indicating the representations don't match.
2. `rep1 != rep2` and `isConversionAllowedBetween(rep1, rep2) == true`. In this case, this function will return false, indicating that we can convert. This is because it is okay to convert from `@convention(c)` with any* cType to `@convention(swift)`.
3. `rep1 != rep2` and `isConversionAllowedBetween(rep1, rep2) == false`. In this case, this function will return true, indicating that we can't convert. This is always correct; e.g. you can't convert from `@convention(swift)` to `@convention(c)`.

Reply:
You can take a look at the test case below in test/Sema/clang_fn_type_mismatch.swift to see the different cases in action.

Reply:
Okay, looks like I did understand it correctly :) I'd really appreciate it if you could add comment about `@conversion(c)` types here for posterity.

Reply:
Added a comment explaining the cases.

Reply:
Thank you!